File: damage.js
"use strict";

function calcComputedSpeed(poke) {
	if (poke.item == "こだわりスカーフ") {
		return int(poke.speed * 1.5);
	}
	if (poke.item == "くろいてっきゅう") {
		return int(poke.speed * 0.5);
	}
	return poke.speed;
}

function calcDamage(userPoke, foePoke, waza) {
	if (isToZero(userPoke, foePoke, waza)) return 0;
	var ch = 1; // 急所はなし
	var r = 85; // 最小乱数

	var type1 = typeChart(waza.type, foePoke.type1);
	var type2 = foePoke.type1 == foePoke.type2 ? 1 : typeChart(waza.type, foePoke.type2);

	// Same Type Attack Bonus
	// ignore てきおうりょく
	var stab = waza.type == userPoke.type1 || waza.type == userPoke.type2 ? 1.5 : 1; 

	// ignore やけど、光の壁、リフレクター、ダブルバトル、雨、晴れ、もらい火
	var mod1 = 1;

	var mod2 = calcMod2(userPoke);
	var mod3 = calcMod3(userPoke, foePoke, waza, (type1*type2) > 1);

	var basePower = calcBasePower(userPoke, foePoke, waza);
	var atk = calcAttack(userPoke, foePoke, waza);
	var def = calcDefence(userPoke, foePoke, waza);
	return calcDamage0(userPoke.level, basePower, atk, def, mod1, ch, mod2, r, stab, type1, type2, mod3);
}

function isToZero(userPoke, foePoke, waza) {
	if (waza.type == "みず" && (foePoke.ability == "ちょすい" || foePoke.ability == "かんそうはだ"))
		return true;
	if (waza.type == "じめん" && foePoke.ability == "ふゆう")
		return true;
	if (waza.type == "ほのお" && foePoke.ability == "もらいび")
		return true;
	if (waza.type == "でんき" && (foePoke.ability == "ちくでん" || foePoke.ability == "でんきエンジン"))
		return true;
	// ignore ぼうおん
	return false;
}

function calcMod2(userPoke) {
	// ignore メトロノーム、さきどり
	if (userPoke.item == "いのちのたま") {
		return 1.3;
	}
	return 1;
}

function calcMod3(userPoke, foePoke, waza, effective) {
	// 一つ掛けるたびに小数点以下切り下げしないといけないので配列を返す

	// ignore いろめがね(効果今ひとつの技が倍になる), ホズのみ(ノーマルタイプ半減)
	var mod3 = [];
	if (foePoke.ability == "ハードロック" && effective) {
		mod3.push(3/4);
	}
	if (userPoke.item == "たつじんのおび" && effective) {
		mod3.push(1.2);
	}
	// 半減実
	if (BERRY_NAMES.indexOf(foePoke.item) == TYPENAMES.indexOf(waza.type) && effective) {
		mod3.push(0.5);
	}
	return mod3;
}

function calcBasePower(userPoke, foePoke, waza, effective) {
	// ignore 威力が固定でない技
	var basePower = waza.basePower;

	// ignore てだすけ(HH), じゅうでん(CHG), どろあそび(MS), みずあそび(WS)

	if (userPoke.item == "ちからのハチマキ" && waza.isPhysical) {
		basePower = int(basePower * 1.1);
	}
	
	if (userPoke.item == "ものしりメガネ" && !waza.isPhysical) {
		basePower = int(basePower * 1.1);
	}

	// ignore プレート, おこう
	// ignore こんごうだま, しらたま, はっきんだま
	
	// ignore とうそうしん, すてみ, てつのこぶし, もうか, しんりょく, げきりゅう, むしのしらせ, テクニシャン

	if (foePoke.ability == "あついしぼう" && (waza.type == "ほのお" || waza.type == "こおり")) {
		basePower = int(basePower * 0.5);
	}
	if (foePoke.ability == "たいねつ" && waza.type == "ほのお") {
		basePower = int(basePower * 0.5);
	}
	if (foePoke.ability == "かんそうはだ" && waza.type == "ほのお") {
		basePower = int(basePower * 1.25);
	}
	return basePower;
}

function calcAttack(userPoke, foePoke, waza) {
	var atk = waza.isPhysical ? userPoke.atk : userPoke.spAtk;

	// ignore 能力変化

	// ignore ヨガパワー ちからもち フラワーギフト こんじょう はりきり スロースタート プラス マイナス ソーラーパワー

	// ignore でんきだま ふといほね こころのしずく しんかいのキバ
	if (waza.isPhysical && userPoke.item == "こだわりハチマキ") {
		atk = int(atk * 1.5);
	}
	if (!waza.isPhysical && userPoke.item == "こだわりメガネ") {
		atk = int(atk * 1.5);
	}
	return atk;
}

function calcDefence(userPoke, foePoke, waza) {
	var def = waza.isPhysical ?  foePoke.def :  foePoke.spDef;
	
	// ignore 能力変化
	// ignore じばく だいばくはつ
	// ignore メタルパウダー ふしぎなうろこ こころのしずく しんかいのウロコ
	// ignore 砂嵐下での岩タイプの特防アップ フラワーギフト

	return def;
}

var TYPENAMES = ["ノーマル", "かくとう", "ひこう", "どく", "じめん", "いわ", "むし", "ゴースト", "はがね", "ほのお", "みず", "くさ", "でんき", "エスパー", "こおり", "ドラゴン", "あく"];

var BERRY_NAMES = ["ホズのみ", "ヨプのみ", "バコウのみ", "ビアーのみ", "シュカのみ", "ヨロギのみ", "タンガのみ", "カシブのみ", "リリバのみ", "オッカのみ", "イトケのみ", "リンドのみ", "ソクノのみ", "ウタンのみ", "ヤチェのみ", "ハバンのみ", "ナモのみ"];

var TYPE_CHART = [
	[1,   1,   1,   1,   1, 0.5,   1,   0, 0.5,   1,   1,   1,   1,   1,   1,   1,   1],
	[2,   1, 0.5, 0.5,   1,   2, 0.5,   0,   2,   1,   1,   1,   1, 0.5,   2,   1,   2],
	[1,   2,   1,   1,   1, 0.5,   2,   1, 0.5,   1,   1,   2, 0.5,   1,   1,   1,   1],
	[1,   1,   1, 0.5, 0.5, 0.5,   1, 0.5,   0,   1,   1,   2,   1,   1,   1,   1,   1],
	[1,   1,   0,   2,   1,   2, 0.5,   1,   2,   2,   1, 0.5,   2,   1,   1,   1,   1],
	[1, 0.5,   2,   1, 0.5,   1,   2,   1, 0.5,   2,   1,   1,   1,   1,   2,   1,   1],
	[1, 0.5, 0.5, 0.5,   1,   1,   1, 0.5, 0.5, 0.5,   1,   2,   1,   2,   1,   1,   2],
	[0,   1,   1,   1,   1,   1,   1,   2,   1,   1,   1,   1,   1,   2,   1,   1, 0.5],
	[1,   1,   1,   1,   1,   2,   1,   1, 0.5, 0.5, 0.5,   1, 0.5,   1,   2,   1,   1],
	[1,   1,   1,   1,   1, 0.5,   2,   1,   2, 0.5, 0.5,   2,   1,   1,   2, 0.5,   1],
	[1,   1,   1,   1,   2,   2,   1,   1,   1,   2, 0.5, 0.5,   1,   1,   1, 0.5,   1],
	[1,   1, 0.5, 0.5,   2,   2, 0.5,   1, 0.5, 0.5,   2, 0.5,   1,   1,   1, 0.5,   1],
	[1,   1,   2,   1,   0,   1,   1,   1,   1,   1,   2, 0.5, 0.5,   1,   1, 0.5,   1],
	[1,   2,   1,   2,   1,   1,   1,   1, 0.5,   1,   1,   1,   1, 0.5,   1,   1,   0],
	[1,   1,   2,   1,   2,   1,   1,   1, 0.5, 0.5, 0.5,   2,   1,   1, 0.5,   2,   1],
	[1,   1,   1,   1,   1,   1,   1,   1, 0.5,   1,   1,   1,   1,   1,   1,   2,   1],
	[1, 0.5,   1,   1,   1,   1,   1,   2,   1,   1,   1,   1,   1,   2,   1,   1, 0.5],
];

function typeChart(wazaType, foeType) {
	return TYPE_CHART[TYPENAMES.indexOf(wazaType)][TYPENAMES.indexOf(foeType)];
}

var int = Math.floor;

function calcDamage0(level, basePower, atk, def, mod1, ch, mod2, r, stab, type1, type2, mod3) {
	var x1 = int((int(level * 2 / 5) + 2) * basePower * atk / 50);
	var x2 = int(int(x1 / def) * mod1) + 2;
	var x3 = int(int(x2 * ch * mod2) * r / 100);
	var x4 = int(int(int(x3 * stab) * type1) * type2);
	mod3.forEach (function (m) {
		x4 = int(x4 * m);
	});
	return x4;
}

function damageTest() {
	console.log(calcDamage0(47, 60, 140, 77, 1, 1, 1, 85, 1.5, 2, 1, []));
	console.log(calcDamage({level: 47, atk: 140, type1: "ノーマル", type2: "ひこう"}, {def: 77, type1: "くさ", type2: "どく"}, {basePower: 60, isPhysical: true, type: "ひこう"}));
}

File: damage.cpp
#include <map>
#include <iostream>
#include <random>
#include "damage.h"
#include "utils.h"
#include "factory.h"

// ref: http://www.smogon.com/dp/articles/damage_formula

const double SM_multiplier[] = { 2. / 8, 2. / 7, 2. / 6, 2. / 5, 2. / 4, 2. / 3, 2. / 2, 3. / 2, 4. / 2, 5. / 2, 6. / 2, 7. / 2, 8. / 2 };

int calcComputedSpeed(PokeStruct * poke) {
	int speed = poke->speed;
	speed *= SM_multiplier[poke->speedMod + 6];
	if (poke->statusCond == STATUSCOND_PARALYSIS) {
		speed *= 0.25;
	}
	if (poke->item == ITEM_こだわりスカーフ) {
		speed *= 1.5;
	}
	if (poke->item == ITEM_くろいてっきゅう) {
		speed *= 0.5;
	}
	return speed;
}

bool isToZero(PokeStruct * userPoke, PokeStruct * foePoke, Waza waza) {
	if (waza->type == TYPE_みず && (foePoke->ability == ABIL_ちょすい || foePoke->ability == ABIL_かんそうはだ))
		return true;
	if (waza->type == TYPE_じめん && foePoke->ability == ABIL_ふゆう)
		return true;
	if (waza->type == TYPE_ほのお && foePoke->ability == ABIL_もらいび)
		return true;
	if (waza->type == TYPE_でんき && (foePoke->ability == ABIL_ちくでん || foePoke->ability == ABIL_でんきエンジン))
		return true;
	// ignore ぼうおん
	return false;
}

double calcMod2(PokeStruct * userPoke) {
	// ignore メトロノーム、さきどり
	if (userPoke->item == ITEM_いのちのたま) {
		return 1.3;
	}
	return 1;
}

map<int, int> BERRY_MAP = {
	{ ITEM_ホズのみ,  0},
	{ ITEM_ヨプのみ, 1 },
	{ ITEM_バコウのみ, 2 },
	{ ITEM_ビアーのみ, 3 },
	{ ITEM_シュカのみ, 4 },
	{ ITEM_ヨロギのみ, 5 },
	{ ITEM_タンガのみ, 6 },
	{ ITEM_カシブのみ, 7 },
	{ ITEM_リリバのみ, 8 },
	{ ITEM_オッカのみ, 10 },
	{ ITEM_イトケのみ, 11 },
	{ ITEM_リンドのみ, 12 },
	{ ITEM_ソクノのみ, 13 },
	{ ITEM_ウタンのみ, 14 },
	{ ITEM_ヤチェのみ, 15 },
	{ ITEM_ハバンのみ, 16 },
	{ ITEM_ナモのみ, 17 },
};


double calcMod3(PokeStruct *userPoke, PokeStruct * foePoke, Waza waza, bool effective) {
	// ignore いろめがね(効果今ひとつの技が倍になる), ホズのみ(ノーマルタイプ半減)
	double mod3 = 1;
	if (foePoke->ability == ABIL_ハ－ドロック && effective) {
		mod3 *= 3.0 / 4;
	}
	if (userPoke->item == ITEM_たつじんのおび && effective) {
		mod3 *= 1.2;
	}
	// 半減実
	if (BERRY_MAP.find(foePoke->item) != BERRY_MAP.end() && BERRY_MAP[foePoke->item] == waza->type && effective) {
		mod3 *= 0.5;
	}
	return mod3;
}

int calcBasePower(PokeStruct *userPoke, PokeStruct *foePoke, Waza waza) {
	// ignore 威力が固定でない技
	int basePower = waza->power;
	if (basePower == 1) {
		throw string("unsupported waza:") + waza->name;
	}

	// ignore てだすけ(HH), じゅうでん(CHG), どろあそび(MS), みずあそび(WS)

	if (userPoke->item == ITEM_ちからのハチマキ && waza->isPhysical) {
		basePower = basePower * 1.1;
	}

	if (userPoke->item == ITEM_ものしりメガネ && !waza->isPhysical) {
		basePower = basePower * 1.1;
	}

	// ignore プレート, おこう
	// ignore こんごうだま, しらたま, はっきんだま

	// ignore とうそうしん, すてみ, てつのこぶし, もうか, しんりょく, げきりゅう, むしのしらせ, テクニシャン

	if (foePoke->ability == ABIL_あついしぼう && (waza->type == TYPE_ほのお || waza->type == TYPE_こおり)) {
		basePower = basePower * 0.5;
	}
	if (foePoke->ability == ABIL_たいねつ && waza->type == TYPE_ほのお) {
		basePower = basePower * 0.5;
	}
	if (foePoke->ability == ABIL_かんそうはだ && waza->type == TYPE_ほのお) {
		basePower = basePower * 1.25;
	}
	if (userPoke->statusCond && waza->no() == WAZA_からげんき) {
		basePower *= 2;
	}
	return basePower;
}

int calcAttack(PokeStruct *userPoke, PokeStruct *foePoke, Waza waza) {
	int atk = waza->isPhysical ? userPoke->atk : userPoke->spAtk;

	if (waza->isPhysical) {
		atk = atk * SM_multiplier[userPoke->atkMod + 6];
	}
	else {
		atk = atk * SM_multiplier[userPoke->spAtkMod + 6];
	}

	// ignore ヨガパワー ちからもち フラワーギフト こんじょう はりきり スロースタート プラス マイナス ソーラーパワー

	// ignore でんきだま ふといほね こころのしずく しんかいのキバ
	if (waza->isPhysical && userPoke->item == ITEM_こだわりハチマキ) {
		atk = atk * 1.5;
	}
	if (!waza->isPhysical && userPoke->item == ITEM_こだわりメガネ) {
		atk = atk * 1.5;
	}
	if (waza->isPhysical && userPoke->ability == ABIL_こんじょう && userPoke->statusCond) {
		atk = atk * 1.5;
	}
	return atk;
}


int calcDefence(PokeStruct * userPoke, PokeStruct * foePoke, Waza waza) {
	int def = waza->isPhysical ? foePoke->def : foePoke->spDef;

	if (waza->isPhysical) {
		def = def * SM_multiplier[foePoke->defMod + 6];
	}
	else {
		def = def * SM_multiplier[foePoke->spDefMod + 6];
	}

	// ignore じばく だいばくはつ
	// ignore メタルパウダー ふしぎなうろこ こころのしずく しんかいのウロコ
	// ignore 砂嵐下での岩タイプの特防アップ フラワーギフト

	return def;
}

int calcDamage0(int level, int basePower, int atk, int def, double mod1, double ch, double mod2, int r,	double stab, double type1, double type2, double mod3) {
	int x1 = (((level * 2 / 5) + 2) * basePower * atk / 50);
	int x2 = ((x1 / def) * mod1) + 2;
	int x3 = ((x2 * ch * mod2) * r / 100);
	int x4 = (((x3 * stab) * type1) * type2);
	int x5 = x4 * mod3;
	return x5;
}

int calcDamage(PokeStruct *userPoke, PokeStruct *foePoke, Waza waza, mt19937 &rnd) {
	if (waza->effectCode == 87) {
		return userPoke->level; // ナイトヘッドとちきゅうなげ
	}
	if (waza->effectCode == 38) {
		return foePoke->maxHp;
	}
	if (waza->power == 0) return 0;
	if (isToZero(userPoke, foePoke, waza)) return 0;
	double ch = 1;
	int r = 85 + rnd() % 16;
	
	if (rnd() % 16 == 0) {
		ch = 2;
		if (userPoke->ability == ABIL_スナイパ－) ch = 3;
	}
	double type1 = TYPE_CHART[waza->type][foePoke->type1];
	double type2 = foePoke->type1 == foePoke->type2 ? 1 : TYPE_CHART[waza->type][foePoke->type2];

	// Same Type Attack Bonus
	int stab = 1;
	if (waza->type == userPoke->type1 || waza->type == userPoke->type2) {
		if (userPoke->ability == ABIL_てきおうりょく) {
			stab = 2;
		}
		else {
			stab = 1.5;
		}
	}

	// ignore やけど、光の壁、リフレクター、ダブルバトル、雨、晴れ、もらい火
	double mod1 = 1;

	double mod2 = calcMod2(userPoke);
	double mod3 = calcMod3(userPoke, foePoke, waza, (type1*type2) > 1);

	int basePower = calcBasePower(userPoke, foePoke, waza);
	int atk = calcAttack(userPoke, foePoke, waza);
	int def = calcDefence(userPoke, foePoke, waza);
	int dmg = calcDamage0(userPoke->level, basePower, atk, def, mod1, ch, mod2, r, stab, type1, type2, mod3);
	return dmg;
}

int main() {
	mt19937 rnd(0);
	PokeStruct poke1 = gen_poke(starter_rank(false, 1), &ENTRIES[1], 0);
	PokeStruct poke2 = gen_poke(starter_rank(false, 1), &ENTRIES[2], 0);
	for (int i = 0; i < 200; i++) {
		cout << calcDamage(&poke1, &poke2, poke1.entry->getWaza(0), rnd) << endl;
	}
	return 0;
}